File: client/lib/utils/media_attachments.dart
enum MediaKind { image, video, pdf, file }

class MediaAttachment {
  final String url;
  final MediaKind kind;
  final String filename;

  const MediaAttachment({
    required this.url,
    required this.kind,
    required this.filename,
  });
}

const Set<String> _imageExtensions = {
  '.png',
  '.jpg',
  '.jpeg',
  '.gif',
  '.webp',
  '.bmp',
  '.svg',
};

const Set<String> _videoExtensions = {
  '.mp4',
  '.webm',
  '.mov',
  '.mkv',
  '.avi',
  '.m4v',
};

const Set<String> _pdfExtensions = {'.pdf'};

String _extensionFromUrl(String rawUrl) {
  final withoutQuery = rawUrl.split('?').first.split('#').first;
  final index = withoutQuery.lastIndexOf('.');
  if (index < 0) return '';
  return withoutQuery.substring(index).toLowerCase();
}

String filenameFromUrl(String rawUrl) {
  final withoutQuery = rawUrl.split('?').first.split('#').first;
  final parts = withoutQuery.split('/').where((part) => part.isNotEmpty).toList();
  if (parts.isEmpty) return 'download';
  return Uri.decodeComponent(parts.last);
}

MediaKind? classifyMediaUrl(String rawUrl) {
  final ext = _extensionFromUrl(rawUrl);
  if (ext.isEmpty) {
    if (rawUrl.contains('/comfyui/outputs/')) return MediaKind.image;
    return null;
  }
  if (_pdfExtensions.contains(ext)) return MediaKind.pdf;
  if (_videoExtensions.contains(ext)) return MediaKind.video;
  if (_imageExtensions.contains(ext)) return MediaKind.image;
  return null;
}

String normalizeMediaUrl(String raw, {String? baseUrl}) {
  var trimmed = raw.trim();
  if (trimmed.startsWith('"') || trimmed.startsWith("'")) {
    trimmed = trimmed.substring(1);
  }
  if (trimmed.endsWith('"') || trimmed.endsWith("'")) {
    trimmed = trimmed.substring(0, trimmed.length - 1);
  }
  if (trimmed.isEmpty) return trimmed;
  if (trimmed.startsWith('http://') || trimmed.startsWith('https://')) return trimmed;
  if (trimmed.startsWith('/')) {
    if (baseUrl == null || baseUrl.isEmpty) return trimmed;
    return '$baseUrl$trimmed';
  }
  if (trimmed.startsWith('workspace/')) {
    final path = '/workspace/download/${trimmed.substring('workspace/'.length)}';
    if (baseUrl == null || baseUrl.isEmpty) return path;
    return '$baseUrl$path';
  }
  return trimmed;
}

void _addAttachment(Map<String, MediaAttachment> found, String rawUrl, {String? baseUrl}) {
  final url = normalizeMediaUrl(rawUrl, baseUrl: baseUrl);
  if (url.isEmpty) return;
  final kind = classifyMediaUrl(url);
  if (kind == null) return;
  if (found.containsKey(url)) return;
  found[url] = MediaAttachment(url: url, kind: kind, filename: filenameFromUrl(url));
}

List<MediaAttachment> extractMediaAttachments(String text, {String? baseUrl}) {
  if (text.trim().isEmpty) return [];
  final found = <String, MediaAttachment>{};

  for (final RegExpMatch match in RegExp(r'\(url:\s*([^)\s]+)\)', caseSensitive: false).allMatches(text)) {
    _addAttachment(found, match.group(1) ?? '', baseUrl: baseUrl);
  }
  for (final RegExpMatch match in RegExp(r'!\[[^\]]*\]\(([^)]+)\)').allMatches(text)) {
    _addAttachment(found, match.group(1) ?? '', baseUrl: baseUrl);
  }
  for (final RegExpMatch match in RegExp(r'\[([^\]]+)\]\(([^)]+)\)').allMatches(text)) {
    _addAttachment(found, match.group(2) ?? '', baseUrl: baseUrl);
  }
  for (final RegExpMatch match in RegExp(r'https?://[^\s<>"]+').allMatches(text)) {
    _addAttachment(found, match.group(0) ?? '', baseUrl: baseUrl);
  }
  for (final RegExpMatch match in RegExp(r'/comfyui/outputs/[^\s<>"]+').allMatches(text)) {
    _addAttachment(found, match.group(0) ?? '', baseUrl: baseUrl);
  }
  for (final RegExpMatch match in RegExp(r'/workspace/download/[^\s<>"]+').allMatches(text)) {
    _addAttachment(found, match.group(0) ?? '', baseUrl: baseUrl);
  }
  for (final RegExpMatch match in RegExp(
    r'workspace/[^\s<>"]+\.(?:png|jpe?g|gif|webp|mp4|webm|mov|pdf)',
    caseSensitive: false,
  ).allMatches(text)) {
    _addAttachment(found, match.group(0) ?? '', baseUrl: baseUrl);
  }

  return found.values.toList();
}

bool isMarkdownImageUrl(String text, String url) {
  return text.contains('![') && text.contains(']($url)');
}

List<MediaAttachment> attachmentsNotInMarkdown(String text, List<MediaAttachment> attachments) {
  return attachments.where((attachment) => !isMarkdownImageUrl(text, attachment.url)).toList();
}

File: src/utils/media-attachments.ts
import * as fsSync from 'fs';
import * as path from 'path';
import { comfyUIService } from '../services/comfyui-service';

export type MediaKind = 'image' | 'video' | 'pdf' | 'file';

export interface MediaAttachment {
  url: string;
  kind: MediaKind;
  filename: string;
}

const IMAGE_EXTENSIONS = new Set(['.png', '.jpg', '.jpeg', '.gif', '.webp', '.bmp', '.svg']);
const VIDEO_EXTENSIONS = new Set(['.mp4', '.webm', '.mov', '.mkv', '.avi', '.m4v']);
const PDF_EXTENSIONS = new Set(['.pdf']);

function extensionFromUrl(rawUrl: string): string {
  const withoutQuery = rawUrl.split('?')[0].split('#')[0];
  return path.extname(withoutQuery).toLowerCase();
}

function filenameFromUrl(rawUrl: string): string {
  const withoutQuery = rawUrl.split('?')[0].split('#')[0];
  const parts = withoutQuery.split('/').filter(Boolean);
  const last = parts[parts.length - 1] ?? 'download';
  try {
    return decodeURIComponent(last);
  } catch {
    return last;
  }
}

export function classifyMediaUrl(rawUrl: string): MediaKind | null {
  const ext = extensionFromUrl(rawUrl);
  if (!ext) {
    if (rawUrl.includes('/comfyui/outputs/')) return 'image';
    return null;
  }
  if (PDF_EXTENSIONS.has(ext)) return 'pdf';
  if (VIDEO_EXTENSIONS.has(ext)) return 'video';
  if (IMAGE_EXTENSIONS.has(ext)) return 'image';
  return null;
}

export function normalizeMediaUrl(raw: string): string {
  const trimmed = raw.trim().replace(/^["']|["']$/g, '');
  if (!trimmed) return trimmed;
  if (trimmed.startsWith('http://') || trimmed.startsWith('https://')) return trimmed;
  if (trimmed.startsWith('/comfyui/') || trimmed.startsWith('/workspace/')) return trimmed;
  if (trimmed.startsWith('workspace/')) {
    return `/workspace/download/${trimmed.slice('workspace/'.length)}`;
  }
  return trimmed;
}

function addAttachment(map: Map<string, MediaAttachment>, rawUrl: string): void {
  const url = normalizeMediaUrl(rawUrl);
  if (!url) return;
  const kind = classifyMediaUrl(url);
  if (!kind) return;
  if (map.has(url)) return;
  map.set(url, { url, kind, filename: filenameFromUrl(url) });
}

export function extractMediaAttachments(text: string): MediaAttachment[] {
  if (!text?.trim()) return [];
  const found = new Map<string, MediaAttachment>();

  for (const match of text.matchAll(/\(url:\s*([^)\s]+)\)/gi)) {
    addAttachment(found, match[1]);
  }
  for (const match of text.matchAll(/!\[[^\]]*]\(([^)]+)\)/g)) {
    addAttachment(found, match[1]);
  }
  for (const match of text.matchAll(/\[([^\]]+)]\(([^)]+)\)/g)) {
    addAttachment(found, match[2]);
  }
  for (const match of text.matchAll(/https?:\/\/[^\s<>"']+/gi)) {
    addAttachment(found, match[0]);
  }
  for (const match of text.matchAll(/\/comfyui\/outputs\/[^\s<>"']+/gi)) {
    addAttachment(found, match[0]);
  }
  for (const match of text.matchAll(/\/workspace\/download\/[^\s<>"']+/gi)) {
    addAttachment(found, match[0]);
  }
  for (const match of text.matchAll(/workspace\/[^\s<>"']+\.(?:png|jpe?g|gif|webp|mp4|webm|mov|pdf)/gi)) {
    addAttachment(found, match[0]);
  }

  return Array.from(found.values());
}

export function resolveMediaAttachmentPath(attachment: MediaAttachment, cwd = process.cwd()): string | null {
  const url = attachment.url;
  const comfyMatch = url.match(/\/comfyui\/outputs\/([^/]+)\/([^/?#]+)/);
  if (comfyMatch) {
    return comfyUIService.resolveOutputFilePath(
      decodeURIComponent(comfyMatch[1]),
      decodeURIComponent(comfyMatch[2]),
    );
  }

  const workspaceDownloadMatch = url.match(/\/workspace\/download\/([^?#]+)/);
  if (workspaceDownloadMatch) {
    const rel = decodeURIComponent(workspaceDownloadMatch[1]);
    return resolveWorkspaceRelativePath(rel, cwd);
  }

  if (url.startsWith('workspace/')) {
    return resolveWorkspaceRelativePath(url.slice('workspace/'.length), cwd);
  }

  if (path.isAbsolute(url) && fsSync.existsSync(url)) {
    return url;
  }

  try {
    const parsed = new URL(url);
    const nestedComfy = parsed.pathname.match(/\/comfyui\/outputs\/([^/]+)\/([^/?#]+)/);
    if (nestedComfy) {
      return comfyUIService.resolveOutputFilePath(
        decodeURIComponent(nestedComfy[1]),
        decodeURIComponent(nestedComfy[2]),
      );
    }
    const nestedWorkspace = parsed.pathname.match(/\/workspace\/download\/([^?#]+)/);
    if (nestedWorkspace) {
      return resolveWorkspaceRelativePath(decodeURIComponent(nestedWorkspace[1]), cwd);
    }
  } catch {
    // not a URL
  }

  return null;
}

function resolveWorkspaceRelativePath(relativePath: string, cwd: string): string | null {
  const workspaceRoot = path.join(cwd, 'workspace');
  const fullPath = path.resolve(workspaceRoot, relativePath);
  if (!fullPath.startsWith(workspaceRoot + path.sep) && fullPath !== workspaceRoot) {
    return null;
  }
  if (!fsSync.existsSync(fullPath)) return null;
  return fullPath;
}
